Doc Updates Only: Add generic view information to plugin development docs. (#1672)
nautobot/nautobot: nautobot/docs/plugins/development.md

This view retrieves a random animal from the database and and passes it as a context variable when rendering a template named `animal.html`, which doesn't exist yet. To create this template, first create a directory named `templates/nautobot_animal_sounds/` within the plugin source directory. (We use the plugin's name as a subdirectory to guard against naming collisions with other plugins.) Then, create a template named `animal.html` as described below.

> ### Utilizing Nautobot Generic Views

Reviewer: Broadly this information is not plugin-specific - I'd personally prefer to see it as a new page under `nautobot/docs/development/`, similar to how we have `nautobot/docs/development/templates.md` describing how to use the generic templates; it could then be linked from here as a cross-reference.

Starting in Nautobot 1.1.0 via [PR](https://github.com/nautobot/nautobot/issues/14), some `generic` views have been exposed to help aid in plugin development.  These views have some requirements that must be in place in order to work.  These can be used by importing them from `from nautobot.core.views import generic`.

Some of the more common views are:

* `ObjectView` - Retrieve a single object for display.
* `ObjectListView` - List a series of objects.
* `ObjectEditView` - Create or edit a single object.
* `ObjectDeleteView` - Delete a single object.
* `BulkCreateView` - Create new objects in bulk.
* `BulkDeleteView` - Delete objects in bulk.

> * `BulkEditView` - Edit objects in bulk.

Reviewer: ```suggestion
* `BulkEditView` - Edit objects in bulk.
* `BulkImportView` - Import objects in bulk from CSV.
```

> Requirements to use the views.
> 
> * Reverse URL naming needs to follow a template of `{modelname}_{method}` where the **model name** is lowercased model class name from `models.py` and **method** is the purpose of the view. E.g. `_list`, `_add`, `_edit`.
> * Specifically for the `ObjectListView` the URLs for each of the `action_buttons` to be utilized must be built out. By default this view has `action_buttons = ("add", "import", "export")`.  This means by default Nautobot would expect to have URLs/Views to be build for `add` and `import`, also `list` since that is the initial URL for this view.
> 
> !!! warning
>     If you're missing any of the aforementioned URLs/Views, when accessing your list view it will result in a error `Reverse for 'None' not found. 'None' is not a valid view function or pattern name.`
> 
> If all of these views are not required you can simply update your view and overload the action buttons.  E.g. `action_buttons = ("add",)` or if none are required `action_buttons = ()`.

Reviewer: ```suggestion
Once you define a view by subclassing any of the above generic classes, you must register it in your `urls.py` as usual. There are a few things to be aware of here:

* Reverse URL naming needs to follow a template of `{modelname}_{method}` where the **model name** is lowercased model class name from `models.py` and **method** is the purpose of the view. E.g. `_list`, `_add`, `_edit`.
* The default rendering context for the `ObjectListView` includes some standard `action_buttons` for interacting with the listed model. By default this view defines `action_buttons = ("add", "import", "export")`. The `export` action is handled automatically by `ObjectListView`, but the `add` and `import` actions need corresponding views in order to work. In other words, if you implement an `ObjectListView` and do not override its `action_buttons`, you must also implement and register the corresponding `ObjectEditView` and `BulkImportView` subclasses as well.

!!! warning
    If you're missing any of the aforementioned URLs/Views, when accessing your list view it will result in a error `Reverse for 'None' not found. 'None' is not a valid view function or pattern name.`

If you do not need `ObjectEditView` and/or `BulkImportView` for your particular model, as an alternative you can simply update your `ObjectListView` subclass to overload the action buttons.  For example, `action_buttons = ("add",)` or if none are required `action_buttons = ()`.
```

> To demonstrate this concepts we can look at the `example_plugin` from the documentation.

Reviewer: ```suggestion
To demonstrate these concepts we can look at the `example_plugin` included in the Nautobot repository.
```

> The example plugin has a simple model called `ExampleModel`.

Reviewer: ```suggestion
The example plugin has a simple model called `ExampleModel`:
```

```python
class ExampleModel(OrganizationalModel):
    name = models.CharField(max_length=20, help_text="The name of this Example.")
    number = models.IntegerField(default=100, help_text="The number of this Example.")

    csv_headers = ["name", "number"]

    class Meta:
        ordering = ["name"]
```

> For the view its inheriting `generic.ObjectListView` and does **not** overload the `action_buttons`.

Reviewer: ```suggestion
The list view for this model subclasses `generic.ObjectListView` and does **not** overload the `action_buttons`:
```

```python
class ExampleModelListView(generic.ObjectListView):
    """List `ExampleModel` objects."""

    queryset = ExampleModel.objects.all()
    filterset = filters.ExampleModelFilterSet
    filterset_form = forms.ExampleModelFilterForm
    table = tables.ExampleModelTable
```

!!! info

>     Since action_buttons was not overloaded, `action_buttons = ("add", "import", "export")` is inherited.

Reviewer: ```suggestion
    Since `action_buttons` was not overloaded, `action_buttons = ("add", "import", "export")` is inherited.
```
